src/charm.py

```diff
@@ -274,14 +292,33 @@ def _create_layer_config(self):
                 "discourse": {
                     "override": "replace",
                     "summary": "Discourse web application",
-                    "command": "sh -c '/srv/scripts/pod_start >>/srv/discourse/discourse.log 2&>1'",
+                    "command": f"sh -c '{SCRIPT_PATH}/pod_start >>/srv/discourse/discourse.log 2&>1'",
                     "startup": "enabled",
                     "environment": self._create_discourse_environment_settings(),
                 }
             },
+            "checks": {
+                "discourse-check": {
+                    "override": "replace",
+                    "http": {"url": f"http://localhost:{SERVICE_PORT}"},
+                },
+            }
         }
         return layer_config
 
+    def _should_run_setup(self):
+        # Return true if setup never ran
+        return not self._stored.setup_ran or (
```

Comment:
Usage of stored state is discouraged https://juju.is/docs/sdk/stored-state-uses-limitations. Is it possible to retrieve the previous value from another source?

Reply:
I discussed this topic with Tom, indeed not the best way.
Unfortunately I don't see any other possibility to know if this script has previously run or not.
Note that the script is "idempotent" so could basically be run for every event linked to this callback, but not sure that's what we want

Reply:
Considering this will be ran only when the config changed event, I'd personally go for running it every time the callback is triggered. Configuration changes are not very frequent so I don't think performance will be an issue here.

Reply:
The name of the method is misleading, it's bound to multiple events:
```
        self.framework.observe(self.on.leader_elected, self._config_changed)
        self.framework.observe(self.on.discourse_pebble_ready, self._config_changed)
        self.framework.observe(self.on.config_changed, self._config_changed)
        self.framework.observe(self.on.upgrade_charm, self._config_changed)
        self.framework.observe(self.on.redis_relation_updated, self._config_changed)
```
Don't know frequency of occurence for these events.

What I could eventually do is write a file on the leader filesystem to store the "state"
But it won't be persisted when the containers is recreated

What do you think ?

Reply:
It seems to me that we can achieve a simpler working solution with some additional refactoring. So I say we park this for now.

From what I see in https://github.com/canonical/discourse-k8s-operator/blob/main/image/scripts/pod_setup, redis is only used to acquire a lock, hence, there's no need to run the hook on `redis_relation_updated`.
Both `discourse_pebble_ready`and `upgrade_charm`won't be triggered without a `config_changed` event, so there's no need for those callbacks, also, I don't see why running it again when leadership changes.

We can get the env var values that really require the execution of the script (probably all used in it) from the containers plan by running `Container.get_plan` and comparing it to the charm config.

Reply:
I had the same thought, however, leadership is only guaranteed for 30s and it can be changed during the hook execution, so I'm not sure removing it is safe.

Reply:
So our discussion gave me an idea, I refactored to use the presence of a pebble plan to launch the startup script, and next to it I use the pebble plan env and compare it with the charm config to know if any value has changed